extensions/amp-powr-player/0.1/amp-powr-player.js

```diff
@@ -0,0 +1,446 @@
+import {VideoEvents} from '../../../src/video-interface';
+import {addParamsToUrl} from '../../../src/url';
+import {
+  createFrameFor,
+  isJsonOrObj,
+  mutedOrUnmutedEvent,
+  objOrParseJson,
+  redispatch,
+} from '../../../src/iframe-video';
+import {dev, user} from '../../../src/log';
+import {dict} from '../../../src/utils/object';
+import {
+  fullscreenEnter,
+  fullscreenExit,
+  getDataParamsFromAttributes,
+  isFullscreenElement,
+  removeElement,
+} from '../../../src/dom';
+import {getData, listen} from '../../../src/event-helper';
+import {
+  installVideoManagerForDoc,
+} from '../../../src/service/video-manager-impl';
+import {isLayoutSizeDefined} from '../../../src/layout';
+
+
+/** @private @const {string} */
+const TAG = 'amp-powr-player';
+
+
+/** @private @const {!Object.<string,string>} */
+const PLAYER_EVENT_MAP = {
+  'ready': VideoEvents.LOAD,
+  'playing': VideoEvents.PLAYING,
+  'pause': VideoEvents.PAUSE,
+  'ended': VideoEvents.ENDED,
+  'ads-ad-started': VideoEvents.AD_START,
+  'ads-ad-ended': VideoEvents.AD_END,
+};
+
+
+/** @implements {../../../src/video-interface.VideoInterface} */
+class AmpPowrPlayer extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Element} */
+    this.iframe_ = null;
+
+    /** @private {?boolean} */
+    this.playing_ = false;
+
+    /** @private {?boolean}  */
+    this.muted_ = false;
+
+    /** @private {?boolean}  */
+    this.frameHasAmpSupport_ = false;
+
+    /** @private {?Promise} */
+    this.playerReadyPromise_ = null;
+
+    /** @private {?Function} */
+    this.playerReadyResolver_ = null;
+
+    /** @private {?number} */
+    this.readyTimeout_ = null;
+
+    /** @private {?Function} */
+    this.unlistenMessage_ = null;
+
+    /**@private {?string} */
+    this.playerId_ = null;
+
+    /** @private {?../../../src/service/url-replacements-impl.UrlReplacements} */
+    this.urlReplacements_ = null;
+  }
+
+  /** @override */
+  preconnectCallback() {
+    this.preconnect.url('https://player.powr.com');
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  viewportCallback(visible) {
+    this.element.dispatchCustomEvent(VideoEvents.VISIBILITY, {
+      visible,
+    });
+  }
+
+  /** @override */
+  buildCallback() {
+    const ampdoc = this.getAmpDoc();
+    const deferred = new Deferred();
+
+    this.urlReplacements_ = Services.urlReplacementsForDoc(ampdoc);
+    this.playerReadyPromise_ = deferred.promise;
+    this.playerReadyResolver_ = deferred.resolve;
+
+    // Warn if the player does not have video interface support
+    this.readyTimeout_ = Services.timerFor(window).delay(() => {
+      user().warn(TAG,
+          'Did not receive ready callback from player %s.' +
+        ' Ensure it has the iframe plugin.', this.playerId_);
+    }, 3000);
+
+    this.playerReadyResolver_(this.iframe_);
```

Comment:
This is resolving the ready promise too early. You wanna resolve inside your `onReady`. Maybe this was meant to be inside the timer above?

Reply:
Good catch. Moved the resolve into `onReady_`.